flusher.go

```diff
@@ -19,57 +19,122 @@ import (
+	s.reportMetricsFlushCounts(ms)
+
+	// we don't report totalHistograms, totalSets, or totalTimers for local veneur instances
+
+	// we cannot do this until we're done using tempMetrics within this function,
+	// since not everything in tempMetrics is safe for sharing
+	go s.flushForward(tempMetrics)
+
+	s.flushRemote(finalMetrics, metricLimit)
+}
+
+type metricsSummary struct {
+	totalCounters   int
+	totalGauges     int
+	totalHistograms int
+	totalSets       int
+	totalTimers     int
+
+	totalLocalHistograms int
+	totalLocalSets       int
+	totalLocalTimers     int
+
+	totalLength int
+}
+
+// tallyMetrics gives a slight overestimate of the number
+// of metrics we'll be reporting, so that we can pre-allocate
+// a slice of the correct length instead of constantly appending
+// for performance
+func (s *Server) tallyMetrics(percentiles []float64) ([]WorkerMetrics, metricsSummary) {
 	// allocating this long array to count up the sizes is cheaper than appending
 	// the []DDMetrics together one at a time
 	tempMetrics := make([]WorkerMetrics, 0, len(s.Workers))
-	var (
-		totalCounters   int
-		totalGauges     int
-		totalHistograms int
-		totalSets       int
-		totalTimers     int
-
-		totalLocalHistograms int
-		totalLocalSets       int
-		totalLocalTimers     int
-	)
+
 	gatherStart := time.Now()
+	ms := metricsSummary{}
+
 	for i, w := range s.Workers {
 		s.logger.WithField("worker", i).Debug("Flushing")
 		wm := w.Flush()
 		tempMetrics = append(tempMetrics, wm)
 
-		totalCounters += len(wm.counters)
-		totalGauges += len(wm.gauges)
-		totalHistograms += len(wm.histograms)
-		totalSets += len(wm.sets)
-		totalTimers += len(wm.timers)
+		ms.totalCounters += len(wm.counters)
+		ms.totalGauges += len(wm.gauges)
+		ms.totalHistograms += len(wm.histograms)
+		ms.totalSets += len(wm.sets)
+		ms.totalTimers += len(wm.timers)
 
-		totalLocalHistograms += len(wm.localHistograms)
-		totalLocalSets += len(wm.localSets)
-		totalLocalTimers += len(wm.localTimers)
+		ms.totalLocalHistograms += len(wm.localHistograms)
+		ms.totalLocalSets += len(wm.localSets)
+		ms.totalLocalTimers += len(wm.localTimers)
 	}
+
 	s.statsd.TimeInMilliseconds("flush.total_duration_ns", float64(time.Now().Sub(gatherStart).Nanoseconds()), []string{"part:gather"}, 1.0)
 
-	totalLength := totalCounters + totalGauges + (totalTimers+totalHistograms)*(HistogramLocalLength+len(percentiles)) +
+	ms.totalLength = ms.totalCounters + ms.totalGauges +
+		// histograms and timers each report a metric point for each percentile
+		// plus a point for their max, min and count
+		(ms.totalTimers+ms.totalHistograms)*(HistogramLocalLength+len(percentiles)) +
 		// local-only histograms will be flushed with percentiles, so we intentionally
-		// use the original percentile list here
-		totalLocalSets + (totalLocalTimers+totalLocalHistograms)*(HistogramLocalLength+len(s.HistogramPercentiles))
-	if s.ForwardAddr == "" {
-		totalLength += totalSets
-	}
+		// use the original percentile list here.
```

Comment:
In the global instance, they'll be `s.HistogramPercentiles`, but in the local instance, they'll be `nil`. In this case, I replaced one but not the other. Thanks for catching!